include/oneapi/dpl/experimental/kt/single_pass_scan.h

```diff
@@ -0,0 +1,245 @@
+
+    static constexpr ::std::size_t __flag_length = sizeof(_StorageType);
+
+    static constexpr _StorageType __partial_mask = 1ul << (__flag_length * 8 - 2);
+    static constexpr _StorageType __full_mask = 1ul << (__flag_length * 8 - 1);
+    static constexpr _StorageType __value_mask = ~(__partial_mask | __full_mask);
+    static constexpr _StorageType __oob_value = __partial_mask | __full_mask;
+
+    static constexpr int __padding = SUBGROUP_SIZE;
+
+    __scan_status_flag(_StorageType* __flags_begin, const std::uint32_t __tile_id)
+        : __atomic_flag(*(__flags_begin + __tile_id + __padding))
+    {
+    }
+
+    void
+    set_partial(const _T __val)
+    {
+        __atomic_flag.store(__val | __partial_mask);
+    }
+
+    void
+    set_full(const _T __val)
+    {
+        __atomic_flag.store(__val | __full_mask);
+    }
+
+    template <typename _Subgroup, typename _BinOp>
+    _T
+    cooperative_lookback(::std::uint32_t __tile_id, const _Subgroup& __subgroup, _BinOp __bin_op,
+                         _StorageType* __flags_begin)
+    {
+        _T __sum{0};
+        auto __local_id = __subgroup.get_local_id();
+
+        for (int __tile = static_cast<int>(__tile_id) - 1; __tile >= 0; __tile -= SUBGROUP_SIZE)
+        {
+            _AtomicRefT __tile_atomic(*(__flags_begin + __tile + __padding - __local_id));
+            _StorageType __tile_val = 0;
+            do
+            {
+                __tile_val = __tile_atomic.load();
+            } while (!sycl::all_of_group(__subgroup, __tile_val != 0));
+
+            bool __is_full = (__tile_val & __full_mask) && ((__tile_val & __partial_mask) == 0);
+            auto __is_full_ballot = sycl::ext::oneapi::group_ballot(__subgroup, __is_full);
+            ::std::uint32_t __is_full_ballot_bits{};
+            __is_full_ballot.extract_bits(__is_full_ballot_bits);
+
+            auto __lowest_item_with_full = sycl::ctz(__is_full_ballot_bits);
+            _T __contribution = __local_id <= __lowest_item_with_full ? __tile_val & __value_mask : _T{0};
+
+            // Sum all of the partial results from the tiles found, as well as the full contribution from the closest tile (if any)
+            __sum += sycl::reduce_over_group(__subgroup, __contribution, __bin_op);
+
+            // If we found a full value, we can stop looking at previous tiles. Otherwise,
+            // keep going through tiles until we either find a full tile or we've completely
+            // recomputed the prefix using partial values
+            if (__is_full_ballot_bits)
+                break;
+        }
+        return __sum;
+    }
+
+    _AtomicRefT __atomic_flag;
+};
+
+template <bool _Inclusive, typename _InRange, typename _OutRange, typename _BinaryOp, typename _KernelParam>
+sycl::event
+single_pass_scan_impl(sycl::queue __queue, _InRange&& __in_rng, _OutRange&& __out_rng, _BinaryOp __binary_op,
+                      _KernelParam)
```

Comment:
How it's already implemented for another algorithm:
```C++
template <bool __is_ascending = true, ::std::uint8_t __radix_bits = 8, typename _KernelParam, typename _KeysIterator>
sycl::event
radix_sort(sycl::queue __q, _KeysIterator __keys_first, _KeysIterator __keys_last, _KernelParam __param = {})
```

So better to have `_KernelParam __param = {}`

Reply:
Thank you for pointing this out. I have changed it to default initialize the kernel param struct.